src/main/scala/TypeEnv.scala

```diff
@@ -0,0 +1,95 @@
+package fuselang
+
+import Syntax._
+import Errors._
+
+object TypeEnv {
+
+  type Stack[T] = List[T]
+  type Scope = Map[Id, Info]
+
+  val emptyEnv: Env = Env(List(Map()))
+
+  case class Env(e: Stack[Scope]) {
+    override def toString = e.foldLeft("")({ case (acc, m) => s"$acc :: $m"})
+    def addScope = Env(Map[Id, Info]() :: e)
+    def endScope = (Env(e.tail), e.head)
+    def apply(id: Id): Info = findBind(e, id) match {
+      case Some(info) => info
+      case None => throw UnboundVar(id)
+    }
+
+    private def findBind(e: Stack[Scope], id: Id): Option[Info] =
```

Comment:
I might call these methods `find` (or `get`), `add`, and `update`, since it's clear that what they work with is bindings.

Reply:
Done.